File: 백준알고리즘/DFS(깊이 우선 탐색)/python/안전 영역.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
graph = []
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
visited = [[False] * N for _ in range(N)]
answer = {}
cnt = 0


for _ in range(N):
    graph.append(list(map(int, input().split())))
M = max(map(max, graph))


def dfs(x, y, n):
    visited[x][y] = True

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if nx < 0 or nx >= N or ny < 0 or ny >= N:
            continue
        if graph[nx][ny] > n and visited[nx][ny] == False:
            dfs(nx, ny, n)


def solution():
    global cnt, visited
    for n in range(0, M + 1):
        for i in range(N):
            for j in range(N):
                if graph[i][j] > n and visited[i][j] == False:
                    cnt += 1
                    dfs(i, j, n)
        visited = [[False] * N for _ in range(N)]
        answer[n] = cnt
        cnt = 0


solution()
print(max(answer.values()))

File: 백준알고리즘/DFS(깊이 우선 탐색)/javascript/안전 영역.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().split('\n');
const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

N = +input();
const graph = [];
for (let i = 0; i < N; i++) {
  graph.push(
    input()
      .split(' ')
      .map(n => +n)
  );
}
const maxV = Math.max.apply(
  null,
  graph.map(n => Math.max.apply(Math, n))
);

let visited = Array.from(Array(N), () => Array(N).fill(0));
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let answer = 0;

function dfs(x, y, base) {
  visited[x][y] = 1;
  const stack = [[x, y]];

  while (stack.length > 0) {
    const cur = stack.pop();

    for (let i = 0; i < 4; i++) {
      const nx = cur[0] + dx[i];
      const ny = cur[1] + dy[i];

      if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;

      if (graph[nx][ny] > base && visited[nx][ny] === 0) {
        visited[nx][ny] = 1;
        stack.push([nx, ny]);
      }
    }
  }
}

for (let k = 0; k <= maxV; k++) {
  let cnt = 0;
  visited = Array.from(Array(N), () => Array(N).fill(0));
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (graph[i][j] > k && visited[i][j] === 0) {
        cnt++;
        dfs(i, j, k);
      }
    }
  }
  answer = Math.max(cnt, answer);
}
console.log(answer);
